File: go/internal/repository/user_repo.go
package repository

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// PostgresUserRepository implements UserRepository with PostgreSQL
type PostgresUserRepository struct {
	pool *pgxpool.Pool
}

// NewPostgresUserRepository creates a new PostgresUserRepository
func NewPostgresUserRepository(pool *pgxpool.Pool) *PostgresUserRepository {
	return &PostgresUserRepository{pool: pool}
}

// Create creates a new user
func (r *PostgresUserRepository) Create(ctx context.Context, user *User) (*User, error) {
	user.ID = uuid.New().String()
	user.CreatedAt = time.Now()
	user.UpdatedAt = time.Now()

	query := `
		INSERT INTO users (id, email, display_name, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5)
		RETURNING id, email, display_name, created_at, updated_at
	`

	err := r.pool.QueryRow(ctx, query,
		user.ID, user.Email, user.DisplayName, user.CreatedAt, user.UpdatedAt,
	).Scan(&user.ID, &user.Email, &user.DisplayName, &user.CreatedAt, &user.UpdatedAt)

	if err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	return user, nil
}

// GetByID retrieves a user by ID
func (r *PostgresUserRepository) GetByID(ctx context.Context, id string) (*User, error) {
	query := `SELECT id, email, display_name, created_at, updated_at FROM users WHERE id = $1`

	user := &User{}
	err := r.pool.QueryRow(ctx, query, id).Scan(
		&user.ID, &user.Email, &user.DisplayName, &user.CreatedAt, &user.UpdatedAt,
	)

	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	return user, nil
}

// Update updates an existing user
func (r *PostgresUserRepository) Update(ctx context.Context, user *User) (*User, error) {
	user.UpdatedAt = time.Now()

	query := `
		UPDATE users 
		SET email = $2, display_name = $3, updated_at = $4
		WHERE id = $1
		RETURNING id, email, display_name, created_at, updated_at
	`

	err := r.pool.QueryRow(ctx, query,
		user.ID, user.Email, user.DisplayName, user.UpdatedAt,
	).Scan(&user.ID, &user.Email, &user.DisplayName, &user.CreatedAt, &user.UpdatedAt)

	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, fmt.Errorf("failed to update user: %w", err)
	}

	return user, nil
}

// Delete deletes a user by ID
func (r *PostgresUserRepository) Delete(ctx context.Context, id string) error {
	query := `DELETE FROM users WHERE id = $1`

	result, err := r.pool.Exec(ctx, query, id)
	if err != nil {
		return fmt.Errorf("failed to delete user: %w", err)
	}

	if result.RowsAffected() == 0 {
		return ErrNotFound
	}

	return nil
}

// List retrieves users with pagination
func (r *PostgresUserRepository) List(ctx context.Context, opts ListOptions) ([]*User, *ListResult, error) {
	if opts.PageSize <= 0 {
		opts.PageSize = 10
	}
	if opts.PageSize > 100 {
		opts.PageSize = 100
	}

	offset := 0
	if opts.PageToken != "" {
		var err error
		offset, err = strconv.Atoi(opts.PageToken)
		if err != nil {
			offset = 0
		}
	}

	// Get total count
	var totalCount int
	err := r.pool.QueryRow(ctx, "SELECT COUNT(*) FROM users").Scan(&totalCount)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to count users: %w", err)
	}

	query := `
		SELECT id, email, display_name, created_at, updated_at 
		FROM users 
		ORDER BY created_at DESC 
		LIMIT $1 OFFSET $2
	`

	rows, err := r.pool.Query(ctx, query, opts.PageSize, offset)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to list users: %w", err)
	}
	defer rows.Close()

	var users []*User
	for rows.Next() {
		user := &User{}
		if err := rows.Scan(&user.ID, &user.Email, &user.DisplayName, &user.CreatedAt, &user.UpdatedAt); err != nil {
			return nil, nil, fmt.Errorf("failed to scan user: %w", err)
		}
		users = append(users, user)
	}

	result := &ListResult{TotalCount: totalCount}
	nextOffset := offset + len(users)
	if nextOffset < totalCount {
		result.NextPageToken = strconv.Itoa(nextOffset)
	}

	return users, result, nil
}

// AddToTenant adds a user to a tenant
func (r *PostgresUserRepository) AddToTenant(ctx context.Context, tenantUser *TenantUser) (*TenantUser, error) {
	if tenantUser.Role == "" {
		tenantUser.Role = "member"
	}
	if tenantUser.Status == "" {
		tenantUser.Status = "active"
	}

	query := `
		INSERT INTO tenant_users (tenant_id, user_id, role, status)
		VALUES ($1, $2, $3, $4)
		ON CONFLICT (tenant_id, user_id) DO UPDATE SET role = $3, status = $4
		RETURNING tenant_id, user_id, role, status
	`

	err := r.pool.QueryRow(ctx, query,
		tenantUser.TenantID, tenantUser.UserID, tenantUser.Role, tenantUser.Status,
	).Scan(&tenantUser.TenantID, &tenantUser.UserID, &tenantUser.Role, &tenantUser.Status)

	if err != nil {
		return nil, fmt.Errorf("failed to add user to tenant: %w", err)
	}

	return tenantUser, nil
}

// RemoveFromTenant removes a user from a tenant
func (r *PostgresUserRepository) RemoveFromTenant(ctx context.Context, tenantID, userID string) error {
	query := `DELETE FROM tenant_users WHERE tenant_id = $1 AND user_id = $2`

	result, err := r.pool.Exec(ctx, query, tenantID, userID)
	if err != nil {
		return fmt.Errorf("failed to remove user from tenant: %w", err)
	}

	if result.RowsAffected() == 0 {
		return ErrNotFound
	}

	return nil
}

// ListTenantUsers lists users in a tenant
func (r *PostgresUserRepository) ListTenantUsers(ctx context.Context, tenantID string, opts ListOptions) ([]*TenantUser, *ListResult, error) {
	if opts.PageSize <= 0 {
		opts.PageSize = 10
	}
	if opts.PageSize > 100 {
		opts.PageSize = 100
	}

	offset := 0
	if opts.PageToken != "" {
		var err error
		offset, err = strconv.Atoi(opts.PageToken)
		if err != nil {
			offset = 0
		}
	}

	// Get total count
	var totalCount int
	err := r.pool.QueryRow(ctx, "SELECT COUNT(*) FROM tenant_users WHERE tenant_id = $1", tenantID).Scan(&totalCount)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to count tenant users: %w", err)
	}

	query := `
		SELECT tenant_id, user_id, role, status 
		FROM tenant_users 
		WHERE tenant_id = $1
		ORDER BY user_id
		LIMIT $2 OFFSET $3
	`

	rows, err := r.pool.Query(ctx, query, tenantID, opts.PageSize, offset)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to list tenant users: %w", err)
	}
	defer rows.Close()

	var tenantUsers []*TenantUser
	for rows.Next() {
		tu := &TenantUser{}
		if err := rows.Scan(&tu.TenantID, &tu.UserID, &tu.Role, &tu.Status); err != nil {
			return nil, nil, fmt.Errorf("failed to scan tenant user: %w", err)
		}
		tenantUsers = append(tenantUsers, tu)
	}

	result := &ListResult{TotalCount: totalCount}
	nextOffset := offset + len(tenantUsers)
	if nextOffset < totalCount {
		result.NextPageToken = strconv.Itoa(nextOffset)
	}

	return tenantUsers, result, nil
}

File: app/repository/user_repo.py
"""
User repository implementation.
"""

import uuid
from datetime import datetime
from typing import List, Tuple

import asyncpg

from app.db.database import Database
from app.repository.models import User, TenantUser, ListOptions, ListResult
from app.repository.errors import NotFoundError


class UserRepository:
    """PostgreSQL user repository."""

    def __init__(self, db: Database):
        self.db = db

    async def create(self, user: User) -> User:
        """Create a new user."""
        user.id = str(uuid.uuid4())
        user.created_at = datetime.now()
        user.updated_at = datetime.now()

        query = """
            INSERT INTO users (id, email, display_name, created_at, updated_at)
            VALUES ($1, $2, $3, $4, $5)
            RETURNING id, email, display_name, created_at, updated_at
        """

        async with self.db.pool.acquire() as conn:
            row = await conn.fetchrow(
                query,
                user.id, user.email, user.display_name, user.created_at, user.updated_at
            )

        return self._row_to_user(row)

    async def get_by_id(self, id: str) -> User:
        """Retrieve a user by ID."""
        query = "SELECT id, email, display_name, created_at, updated_at FROM users WHERE id = $1"

        async with self.db.pool.acquire() as conn:
            row = await conn.fetchrow(query, id)

        if not row:
            raise NotFoundError(f"user not found: {id}")

        return self._row_to_user(row)

    async def update(self, user: User) -> User:
        """Update an existing user."""
        user.updated_at = datetime.now()

        query = """
            UPDATE users 
            SET email = $2, display_name = $3, updated_at = $4
            WHERE id = $1
            RETURNING id, email, display_name, created_at, updated_at
        """

        async with self.db.pool.acquire() as conn:
            row = await conn.fetchrow(
                query,
                user.id, user.email, user.display_name, user.updated_at
            )

        if not row:
            raise NotFoundError(f"user not found: {user.id}")

        return self._row_to_user(row)

    async def delete(self, id: str) -> None:
        """Delete a user by ID."""
        query = "DELETE FROM users WHERE id = $1"

        async with self.db.pool.acquire() as conn:
            result = await conn.execute(query, id)

        if result == "DELETE 0":
            raise NotFoundError(f"user not found: {id}")

    async def list(self, opts: ListOptions) -> Tuple[List[User], ListResult]:
        """Retrieve users with pagination."""
        page_size = max(1, min(opts.page_size or 10, 100))
        offset = 0
        if opts.page_token:
            try:
                offset = int(opts.page_token)
            except ValueError:
                offset = 0

        async with self.db.pool.acquire() as conn:
            total_count = await conn.fetchval("SELECT COUNT(*) FROM users")

            query = """
                SELECT id, email, display_name, created_at, updated_at 
                FROM users 
                ORDER BY created_at DESC 
                LIMIT $1 OFFSET $2
            """
            rows = await conn.fetch(query, page_size, offset)

        users = [self._row_to_user(row) for row in rows]

        result = ListResult(total_count=total_count)
        next_offset = offset + len(users)
        if next_offset < total_count:
            result.next_page_token = str(next_offset)

        return users, result

    async def add_to_tenant(self, tenant_user: TenantUser) -> TenantUser:
        """Add a user to a tenant."""
        if not tenant_user.role:
            tenant_user.role = "member"
        if not tenant_user.status:
            tenant_user.status = "active"

        query = """
            INSERT INTO tenant_users (tenant_id, user_id, role, status)
            VALUES ($1, $2, $3, $4)
            ON CONFLICT (tenant_id, user_id) DO UPDATE SET role = $3, status = $4
            RETURNING tenant_id, user_id, role, status
        """

        async with self.db.pool.acquire() as conn:
            row = await conn.fetchrow(
                query,
                tenant_user.tenant_id, tenant_user.user_id, tenant_user.role, tenant_user.status
            )

        return self._row_to_tenant_user(row)

    async def remove_from_tenant(self, tenant_id: str, user_id: str) -> None:
        """Remove a user from a tenant."""
        query = "DELETE FROM tenant_users WHERE tenant_id = $1 AND user_id = $2"

        async with self.db.pool.acquire() as conn:
            result = await conn.execute(query, tenant_id, user_id)

        if result == "DELETE 0":
            raise NotFoundError(f"tenant_user not found: tenant_id={tenant_id}, user_id={user_id}")

    async def list_tenant_users(self, tenant_id: str, opts: ListOptions) -> Tuple[List[TenantUser], ListResult]:
        """List users in a tenant."""
        page_size = max(1, min(opts.page_size or 10, 100))
        offset = 0
        if opts.page_token:
            try:
                offset = int(opts.page_token)
            except ValueError:
                offset = 0

        async with self.db.pool.acquire() as conn:
            total_count = await conn.fetchval(
                "SELECT COUNT(*) FROM tenant_users WHERE tenant_id = $1",
                tenant_id
            )

            query = """
                SELECT tenant_id, user_id, role, status 
                FROM tenant_users 
                WHERE tenant_id = $1
                ORDER BY user_id
                LIMIT $2 OFFSET $3
            """
            rows = await conn.fetch(query, tenant_id, page_size, offset)

        tenant_users = [self._row_to_tenant_user(row) for row in rows]

        result = ListResult(total_count=total_count)
        next_offset = offset + len(tenant_users)
        if next_offset < total_count:
            result.next_page_token = str(next_offset)

        return tenant_users, result

    def _row_to_user(self, row: asyncpg.Record) -> User:
        """Convert a database row to a User object."""
        return User(
            id=str(row["id"]),
            email=row["email"],
            display_name=row["display_name"],
            created_at=row["created_at"],
            updated_at=row["updated_at"],
        )

    def _row_to_tenant_user(self, row: asyncpg.Record) -> TenantUser:
        """Convert a database row to a TenantUser object."""
        return TenantUser(
            tenant_id=str(row["tenant_id"]),
            user_id=str(row["user_id"]),
            role=row["role"],
            status=row["status"],
        )
